WTFMove() is not supposed to compile when a `const` variable is passed in

cdumez opened the pull request:
#### 021be8e5aefd8742e091c6369be13fd745207359
<pre>
WTFMove() is not supposed to compile when a `const` variable is passed in
<a href="https://bugs.webkit.org/show_bug.cgi?id=302139">https://bugs.webkit.org/show_bug.cgi?id=302139</a>

Reviewed by Ryosuke Niwa and Geoffrey Garen.

WTFMove() is not supposed to compile when a `const` variable is passed in,
it the main reason we use `WTFMove()` over `std::move()`. However, this is
currently not working as intended and it is compiling just fine. This can
lead to bugs such as Bug 301998.

Address the issue by no longer extending the `std` namespace since this is
undefined behavior [1]. Instead define `move()` in the WTF
namespace.

Most of the incorrect uses of WTFMove() were addressed in 302773@main already.

Unfortunately, I wasn&apos;t able to get the `permissive-move` compiler warnings
to work with our custom move function. It seems the compiler hard-codes the
std::move() name so it doesn&apos;t warn with code like this:
```
String foo()
{
    StringBuilder builder;
    String string = builder.toString();
    return WTFMove(string); // Should warn about unnecessary move but doesn&apos;t.
}
```
I tried a lot of things but it seems we either use the default `std::move()`
and get the `permissive-move` warnings OR we use our custom move functions
and then we can add static assertions to catch unnecessary moves and moves
of const values that would lead to silent copying.

[1] <a href="https://en.cppreference.com/w/cpp/language/extending_std.html">https://en.cppreference.com/w/cpp/language/extending_std.html</a>

* Source/WTF/wtf/StdLibExtras.h:
(WTF::move):
(std::move): Deleted.
* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::requestImmersive):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::fromIPCData):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _validateText:inNode:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/304711@main">https://commits.webkit.org/304711@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e817c56f57afae7642ee242fbf50e7e7682a62c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136235 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143945 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5f7f5a4f-6b01-47b0-bee3-fbfb67dfb86e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9267 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8436 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104168 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2a2edf64-89d7-4a16-802b-46b877642a60) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6743 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122078 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84999 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/897f3143-32b8-4a6c-86cd-edceb7d3a47d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6400 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4058 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4539 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128193 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115687 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40279 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146691 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134720 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8275 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40846 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112512 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8291 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112853 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28667 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6326 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118383 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8323 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36440 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167499 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8041 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71882 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43705 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8262 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8115 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -846,23 +846,19 @@ inline void* operator new(size_t, NotNullTag, void* location)
-namespace std {
+namespace WTF {
 
 template<WTF::CheckMoveParameterTag, typename T>
```
I don't think we need `CheckMoveParameter` if we aren't trying to define `move` in the `std` namespace.

cdumez committed:
WTFMove() is not supposed to compile when a `const` variable is passed in
https://bugs.webkit.org/show_bug.cgi?id=302139

Reviewed by Ryosuke Niwa and Geoffrey Garen.

WTFMove() is not supposed to compile when a `const` variable is passed in,
it the main reason we use `WTFMove()` over `std::move()`. However, this is
currently not working as intended and it is compiling just fine. This can
lead to bugs such as Bug 301998.

Address the issue by no longer extending the `std` namespace since this is
undefined behavior [1]. Instead define `move()` in the WTF
namespace.

Most of the incorrect uses of WTFMove() were addressed in 302773@main already.

Unfortunately, I wasn't able to get the `permissive-move` compiler warnings
to work with our custom move function. It seems the compiler hard-codes the
std::move() name so it doesn't warn with code like this:
```
String foo()
{
    StringBuilder builder;
    String string = builder.toString();
    return WTFMove(string); // Should warn about unnecessary move but doesn't.
}
```
I tried a lot of things but it seems we either use the default `std::move()`
and get the `permissive-move` warnings OR we use our custom move functions
and then we can add static assertions to catch unnecessary moves and moves
of const values that would lead to silent copying.

[1] https://en.cppreference.com/w/cpp/language/extending_std.html

* Source/WTF/wtf/StdLibExtras.h:
(WTF::move):
(std::move): Deleted.
* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::requestImmersive):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::fromIPCData):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _validateText:inNode:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/304711@main
